### ISSUE TITLE
Result doesn't not work well with pipelines

### ISSUE THREAD

**Issue by dylandoamaral** [flix/flix#2801]
I am falling in love with the pipeline concept, however it doesn't work well with Result's functions such as:
`pub def getWithDefault(r: Result[t, e], d: t): t`
Wouldn't it be better if the definition was:
`pub def getWithDefault(d: t, r: Result[t, e]): t`

It would allow us to do:
```
File.readLines("./input.txt") |> Result.getWithDefault("") # String & Impure
```

I can try to do it myself if you validate the issue !

**Comment by magnus-madsen** [comment]
Agreed! We should fix this. This also seems to violate our "subject-last" principle.

**Comment by magnus-madsen** [comment]
I think the first thing to do is to compile a list of `Result.X` functions that violate this principle. Once we have them we can discuss how to fix each.

**Comment by magnus-madsen** [comment]
@dylandoamaral And thanks for the bug report and welcome to our little community :)

**Comment by dylandoamaral** [comment]
I will do it when I have a bit more time, thank you !

**Comment by stephentetley** [comment]
`Result.getWithDefault` might be better with its args swapped but you can use `flip` in the meantime to still have a pipeline:

> File.readLines("./input.txt") |> flip(Result.getWithDefault, Nil) |> List.foreach(println);

**Comment by dylandoamaral** [comment]
Here is the list of the definition that doesn't respect the "subject-last" principle:
```
pub def getWithDefault(r: Result[t, e], d: t): t
pub def withDefault(r1: Result[t, e1], r2: Result[t, e2]): Result[t, e2]
pub def traverse(f: a -> Result[b, e] & f, xs: List[a]): Result[List[b], e] & f
pub def traverseX(f: a -> Result[b, e] & f, xs: List[a]): Result[Unit, e] & f
pub def foldLeftM(f: (b, a) -> Result[b, e] & f, s: b, xs: List[a]): Result[b, e] & f
pub def foldRightM(f: (a, b) -> Result[b, e] & f, s: b, xs: List[a]): Result[b, e] & f
pub def liftN(f: (t1, t2) -> u & f, r1: Result[t1, e], r2: Result[t2, e]): Result[u, e] & f /// all lift
```

There is some I am not sure about:
```
def sequenceHelper(xs: List[Result[a, e]], fk: e -> Result[List[a], e], sk: List[a] -> Result[List[a], e]): Result[List[a], e]
def traverseHelper(f: a -> Result[b, e] & f, xs: List[a], fk: e -> Result[List[b], e], sk: List[b] -> Result[List[b], e]): Result[List[b], e] & f
def foldRightMHelper(f: (a, b) -> Result[b, e] & f, s: b, xs: List[a], fk: e -> Result[b, e] & f, sk: b -> Result[b, e] & f): Result[b, e] & f
```

**Comment by dylandoamaral** [comment]
I think there is the same problem in List, for example:
```
pub def zip(xs: List[a], ys: List[b]): List[(a,b)] 
should be
pub def zip(xs: List[b], ys: List[a]): List[(a,b)] 
```
Do you want me to open another ticket ?

**Comment by mlutze** [comment]
We've had some discussion about `zip`. The issue with swapping the parameters is that it might be unintuitive to use the function "normally" (not using `|>`). Perhaps with the new named parameter idiom, we could do something like 
```scala
pub def zip(right: {right :: List[b]}, left: List[a]): List[(a, b)]
```
to maintain the "subject-last" idea.

**Comment by magnus-madsen** [comment]
I looked at `Result` and I think these should be changed:

- [ ] `Result.getWithDefault` 
- [ ] `Result.withDefault`.

I worried about the last one, because there is opportunity for confusion and silent breakage. I think `withDefault` would be a good candidate for having its first argument be a record `{default :: String}`. (e.g. like https://github.com/flix/flix/blob/master/main/src/library/String.flix#L1070)

**Comment by dylandoamaral** [comment]
I will take it if you are ok with it !

**Comment by magnus-madsen** [comment]
Sure. Give it a go!

**Comment by stephentetley** [comment]
I think there will always be a bit of taste or judgement to choosing the argument order - principles help to make a judgement but principles might need to be overruled.

For instance, I don't look at `zip` and see a "subject", I see it as being more like multiple dispatch which was important concept in the Lisp world with both arguments being equally "subjects". That said I've been using various functional languages as my goto language for a long time so this is quite ingrained for me.

These four I see as List traversals - they happen to be in the Result module because they use Result as a (captial I) "Idiom" or Applicative / Monad, but I'd think of the list as subject:

~~~

pub def traverse(f: a -> Result[b, e] & f, xs: List[a]): Result[List[b], e] & f
pub def traverseX(f: a -> Result[b, e] & f, xs: List[a]): Result[Unit, e] & f
pub def foldLeftM(f: (b, a) -> Result[b, e] & f, s: b, xs: List[a]): Result[b, e] & f
pub def foldRightM(f: (a, b) -> Result[b, e] & f, s: b, xs: List[a]): Result[b, e] & f

~~~

The `liftN` functions are their own coding style (i.e. a little "i" idiom). The FParsec library in F# has subject last versions that it calls `pipeN` but I haven't noticed similar elsewhere.

**Comment by dylandoamaral** [comment]
Fine, as requested by @magnus-madsen I will only focus on `Result.getWithDefault` and `Result.withDefault`.

For a pipeline perspective it makes sense to respect the last subject principle such as `1 :: 2 :: Nil |> List.zip(3 :: 4 :: Nil) -> (1, 3) :: (2, 4) :: Nil` however it would also be confusing when we use `List.zip(1 :: 2 :: Nil, 3 :: 4 :: Nil ) -> (3, 1) :: (4, 2) :: Nil` so I don't know which one is the best.

**Comment by dylandoamaral** [comment]
> I looked at `Result` and I think these should be changed:
> 
> * [ ]  `Result.getWithDefault`
> * [ ]  `Result.withDefault`.
> 
> I worried about the last one, because there is opportunity for confusion and silent breakage. I think `withDefault` would be a good candidate for having its first argument be a record `{default :: String}`. (e.g. like https://github.com/flix/flix/blob/master/main/src/library/String.flix#L1070)

Before I close the issue someone can explain to me why we have to use a record here ?

**Comment by magnus-madsen** [comment]
@dylandoamaral There is a bit of discussion here:

https://github.com/flix/flix/issues/2442
https://github.com/flix/flix/issues/2457
https://github.com/flix/flix/pull/2421
https://github.com/flix/flix/issues/2345

**Comment by magnus-madsen** [comment]
We have a problem where we cannot define:

```
def withDefault(r : {default :: } ...
```

because `default` is a keyword.

We need to come up with a new name before we can proceed.

**Comment by dylandoamaral** [comment]
I propose 
```
def orElse(that: {that:: } ...
```

**Comment by magnus-madsen** [comment]
So its `getWithDefault` and `orElse(that)`? What do people think?

**Comment by dylandoamaral** [comment]
No I thought about `orElse` and `getOrElse`

**Comment by magnus-madsen** [comment]
One problem is that `Map` has `getWithDefault` so now we renaming that too?

**Comment by dylandoamaral** [comment]
I guess so, the same apply for Option and so on, however I didn't think that my issue would cause such a headache ^^'. Maybe we have to wait some more inputs from other people.

**Comment by magnus-madsen** [comment]
That's what language (library) design is all about :)

We could just go with `getWithDefault(r: {alt :: a}` for now. Then, if we get rid of `default`, we can fix it 100% later.

**Comment by jaschdoc** [comment]
> We've had some discussion about `zip`. The issue with swapping the parameters is that it might be unintuitive to use the function "normally" (not using `|>`). Perhaps with the new named parameter idiom, we could do something like
> 
> ```scala
> pub def zip(right: {right :: List[b]}, left: List[a]): List[(a, b)]
> ```
> 
> to maintain the "subject-last" idea.

I fear that may cause unwanted confusion for functions that use the same pattern like `append` for instance.
OCaml also uses the leftmost argument as the leftmost element in the tuple (like what we currently have).

https://ocaml.org/api/List.html search for `combine`

**Comment by magnus-madsen** [comment]
We will not change zip nor append. That is too far.
